Extend convenience API for QP + doc (#455)
google/jaxopt: docs/quadratic_programming.rst

  u = targets, jnp.full(data.shape[1], jnp.inf), jnp.zeros_like(c[0])

  hyper_params = dict(params_obj=(None, c), params_eq=data, params_ineq=(l, u))
  osqp = BoxOSQP(matvec_Q=matvec_Q, matvec_A=matvec_A, tol=1e-2)
  sol, state = osqp.run(None, **hyper_params)

Quadratic Polynomial API
------------------------

Solvers :class:`EqualityConstrainedQP`, :class:`OSQP` and :class:`BoxOSQP` support the quadratic polynomial API.
It means that the whole objective function `x -> 1/2 x^T Q x + c^T x + K` can be provided as a function
``fun`` that computes the quadratic polynomial. The function must be differentiable w.r.t `x`.

It offers several advantages:
  * the code is easier to read and closer to the mathematical formulation of the problem.
  * there is no need to provide the matrix `Q` and the vector `c` separately, nor to remove the constant term `K`.
  * the derivatives w.r.t (params_obj, params_eq, params_ineq) may be even easier to compute
    than materializing the full matrix.

Take care that this API also have drawbacks:
  * the function ``fun`` must be differentiable w.r.t `x` (with Jax's AD), even if you are not interested in the derivatives of your QP.

>   * the derivation required to extract `x -> Qx` and `c` may be expensive and slow.

Reviewer: ```suggestion
  * to extract `x -> Qx` and `c` from the function, we need to compute the Hessian-vector product and the gradient of ``fun``, which may be expensive.
```

>   * the shape of primal variables `x` cannot be inferred from `c`, so the user must provide it.

Reviewer: ```suggestion
  * for this API `init_params` must be provided to `run`, contrary to the other APIs.
```

> We illustrate with Non Negative Least Squares (NNLS) factorization::

Reviewer: ```suggestion
We illustrate this API with Non Negative Least Squares (NNLS)::
```

  W_sol = solver.run(init_params=init_params, **hyper_params).params.primal

This API is not recommended for large-scale problems or nested differentiations, use matvec API instead.

Implicit differentiation pitfalls
---------------------------------

When using implicit differentiation, the parameters w.r.t which we differentiate
must be passed to `params_obj`, `params_eq` or `params_ineq`. They should not be captured
from the global scope by `fun` or `matvec`. We illustrate below this common mistake::

  def wrong_solver(Q):  # don't do this!

    def matvec_Q(params_Q, x):
      del params_Q  # unused
      # error! Q is captured from the global scope.
      # it does not fail now, but it will fail later.
      return jnp.dot(Q, x)
    
    c = jnp.zeros(Q.shape[0])

    A = jnp.array([[1.0, 2.0]])
    b = jnp.array([1.0])

    eq_qp = EqualityConstrainedQP(matvec_Q=matvec_Q)
    sol = eq_qp.run(None, params_obj=(None, c), params_eq=(A, b)).params
    loss = jnp.sum(sol.primal)
    return loss

  Q = jnp.array([[1.0, 0.5], [0.5, 4.0]])
  _ = wrong_solver(Q)  # no error... but it will fail later.
  _ = jax.grad(wrong_solver)(Q)  # raise CustomVJPException

> Also, notice that since the problems are convex the optimum is independant of the

Reviewer: ```suggestion
Also, notice that since the problems are convex, the optimum is independent of the
```

starting point `init_params`. Hence, derivatives w.r.t `init_params` are always

> zero (mathematically), but will yield a failure in implicit differentiation,

Reviewer: ```suggestion
zero (mathematically).
```

> or garbage results in case of unrolling (for iterative solvers).  

Reviewer: Remove
